lib/timers/promises.js

```diff
@@ -122,7 +124,110 @@ function setImmediate(value, options = {}) {
   });
 }
 
+function setInterval(after, value, options = {}) {
+  const args = value !== undefined ? [value] : value;
+  if (options == null || typeof options !== 'object') {
+    throw new ERR_INVALID_ARG_TYPE(
+      'options',
+      'Object',
+      options);
+  }
+  const { signal, ref = true } = options;
+  if (signal !== undefined &&
+    (signal === null ||
+      typeof signal !== 'object' ||
+      !('aborted' in signal))) {
+    throw new ERR_INVALID_ARG_TYPE(
+      'options.signal',
+      'AbortSignal',
+      signal);
+  }
+  if (typeof ref !== 'boolean') {
+    throw new ERR_INVALID_ARG_TYPE(
+      'options.ref',
+      'boolean',
+      ref);
+  }
+  return {
+    [Symbol.asyncIterator]() {
+      let timeout,
+        deferred;
+      let active = true;
+      const iterator = {
+        next() {
```

Comment:
this timing logic is incorrect. some examples:

```js
const it = setInterval(n);
it.next(); // should resolve after n seconds
it.next(); // should resolve after n*2 seconds
```

```js
const it = setInterval(n);
await delay(n / 2);
it.next(); // should resolve after n/2 seconds
it next(); // should resolve after (n/2)+n seconds
```

Reply:
This may have to be an option to enable one or the other timing.. 

Here is an example that brings the idea behind using timing for each `.next()` 

```js
const iterable = setInterval(100);
const iterator = iterable[Symbol.asyncIterator]();
await iterator.next()
await delay(150)
await iterator.next()
await delay(150)
```

This would be the same as

```
for await (const unused of setInterval(100)) {
  await delay(150)
}
```

There is also a difference between the "async iterable" instance (that can be passed directly to `for await`) vs the iterator creation time, and then the next request to wait for the next interval

```js
const iterable = setInterval(100);
const iterator = iterable[Symbol.asyncIterator]();
const first = iterator.next()
const second = iterator.next()
```

The above code seems a bit confusing... we would need to "queue" those resolves then, maybe thats fine? 

```js
const defers = []

setInterval(() => {
  const next = defers.shift()
  next.resolve({ done: false, value })
}, after)

/* ... */

next() {
  /* ... */
  
  const deferred = /* ... */
  defers.push(deferred)
  return deferred.promise
},
return() {
   /* ... */
   
   const result = { done: true, value: undefined }
   defers.forEach(deferred => deferred.resolve(result))
   return result
}
```

-------------------

I think one, or the other, or both sets of logic will be good, just need to know what one

Reply:
Im in favour of having:

- Timeouts triggered within `next` as the default behaviour (keeping consistent behaviour across all use cases)
- A single interval for all iterators as an option
- A single interval for each iterator as an option (as it would be if we implemented like https://github.com/nodejs/node/pull/35841#issuecomment-717782354)

If we add the two additional options, this would definitely add to the overall complexity for sure, I would think we would need three completely seperate implementations for this. 

----

I believe `next` should be asserted to being called only after the previous promise was resolved as well (e.g. `deferred` not existing in the current implementation), rather than leaving this open to interpretation, else we will most likely need to queue multiple promises to be resolved for each iterator at the same time, rather than one in flight per iterator

Reply:
> A single interval for each iterator as an option (as it would be if we implemented like #35841 (comment))

Isn't that the mental model when someone does `setInterval`? 

The disadvantage of my implementation in that comment is "what happens if the user 'misses' an interval?" The timing is right but the code should probably push the callbacks to an array (instead of replacing a reference) and fulfil them in the interval.

Also cc @ronag wdyt about what we should do with timers -> async iterators conversion in terms of behavior?

Reply:
Reading the HTML spec for `setInterval`, it runs the `setTimeout` steps, with a "repeat" flag that repeats the steps again post task (which for the `setInterval`, it is a sync task)

https://html.spec.whatwg.org/multipage/timers-and-user-prompts.html#dom-setinterval

> The setInterval() method must return the value returned by the **timer initialization steps**, passing them the method's arguments, the object on which the method for which the algorithm is running is implemented (a Window or WorkerGlobalScope object) as the method context, and the repeat flag set to true.

The repeating happens _after_ the related tasks function has been invoked (which in our case is the external code executing, the stuff between `next` calls) and for any errors to be caught 

> Run the appropriate set of steps from the following list:
> 
> If the first method argument is a Function
Invoke the Function. Use the third and subsequent method arguments (if any) as the arguments for invoking the Function. Use method context proxy as the callback this value. If this throws an exception, catch it, and report the exception.

Then 

> If the repeat flag is true, then call timer initialization steps again, passing them the same method arguments, the same method context, with the repeat flag still set to true, and with the previous handle set to handler.

The running of the task now is async, we tell the consumer to start the task by resolving the task, and the consumer signifies the task is complete by calling `next` again (or `return`) 

-------------------

Aside from the above, we could warn if we skip intervals and leave it at that, but then this seems like inconsistent behaviour, if I had a 30 minute interval going, happened to take 31 minutes for my task, and now have to wait for 29 minutes till the next interval.... leading to the intervals actually being once an hour

Reply:
Conceptually I would imagine it to be equivalent to adding an `await` on `timer._onTimeout()` here:

https://github.com/nodejs/node/blob/bec918fb9bbf3fa1d8f9f5fcd37a9ee5f719dd63/lib/internal/timers.js#L550-L562

Reply:
Ah! As I say that, I see `start` is set _before_ `_onTimeout` is invoked 🤔

Reply:
> Reading the HTML spec for setInterval, it runs the setTimeout steps, with a "repeat" flag that repeats the steps again post task (which for the setInterval, it is a sync task)

Well, two things:
 - That's just an easy way to specify things and doesn't actually indicate when it runs.
 - We don't follow that (the web timers) spec (even remotely) though if we built Node today then we _probably_ would.

What I meant is just "the standard queueing trick" something like:

```js
async function* iteratorSetInterval(ms, options = {}) {
  // validateOptions
  // validateAbortSignal
  let callbacks = [], pending = 0;

  let aborted = options && options.signal && options.signal.aborted;
  if (options && options.signal) {
    options.signal.addEventListener('abort', () => {
      aborted = true;
    });
  }
  
  const timerRef = setInterval(() => {
    if (callbacks.length > 0) callbacks.shift()();
    else pending++;
  }, ms);
  if (options && options.ref === false) {
    timerRef.unref();
  }

  try {
    while (!aborted) {
      yield await new Promise((resolve) => { 
         if (pending > 0) { resolve(); pending--; } ;
         else callbacks.push(resolve);
      });
    }
  } finally {
    clearInterval(timerRef);
  }
}
```